### ISSUE TITLE
Query condition `equalTo` doesn't work for relation field with `Parse.User`

### ISSUE THREAD

**Issue by stephannielsen** [parse-community/Parse-SDK-JS#1534]
### New Issue Checklist

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am not just asking a [question](https://github.com/parse-community/.github/blob/master/SUPPORT.md).
- [x] I have searched through [existing issues](https://github.com/parse-community/Parse-SDK-JS/issues?q=is%3Aissue).
- [x] I can reproduce the issue with the latest versions of [Parse Server](https://github.com/parse-community/parse-server/releases) and the [Parse JS SDK](https://github.com/parse-community/Parse-SDK-JS/releases). 

### Issue Description

In our Cloud functions, we use `request.user` in a query to filter the results on a `Parse.Relation` field.

```
// class schema:
export const MyClassSchema = {
  className: 'MyClass',
  fields: {
    name: { type: 'String', required: true },
    users: { type: 'Relation', targetClass: '_User' },
  },
  classLevelPermissions: {
    find: { requiresAuthentication: true },
    count: { requiresAuthentication: true },
    get: { requiresAuthentication: true },
    update: {},
    create: {},
    delete: {},
    protectedFields: {
      '*': ['users'],
    },
  },
};
```

```
// in a cloud function which requires user authentication
    const query = new Parse.Query('MyClass')
      .equalTo('users', request.user);
    
    const result = await query.find(useMasterKeyOption);
```

Since 3.4.2, this returns an empty list, before it returned the correct list of objects which have `request.user` in their relation.
However, if I create a new user User pointer with `request.user.id`, the lookup works fine again.

```
    const user = new Parse.User();
    user.id = request.user.id;
    const query = new Parse.Query('MyClass')
      .equalTo('users', user);
    const result = await query.find(useMasterKeyOption);
```

Another workaround is:
```
const query = new Parse.Query('MyClass').equalTo('users', request.user.toPointer());
```

`request.user` looks fine in the debugger. I also looked at the release notes between 3.4.1 and 3.4.2 but nothing catched my eye.

### Steps to reproduce

Use a class schema with relation field to Parse.User and test with a cloud function within a user scope.

### Actual Outcome

Empty result set.

### Expected Outcome

Result set with entries where user is in the relation.

###  Environment

Server
- Parse Server version: `5.2.4`
- Operating system: `macOS 12.4`
- Local host: Node 16 LTS Docker container
- TypeScript `4.6.4`

Database
- System (MongoDB or Postgres): `MongoDB`
- Database version: `5.0.8`
- Local or remote host (MongoDB Atlas, mLab, AWS, Azure, Google Cloud, etc): `Local`

Client
- Parse JS SDK version: `3.4.2`, `3.4.3`, `3.4.4`

**Comment by stephannielsen** [comment]
I also see an error with user pointer fields:

```js
    const userQuery = new Parse.Query(
      'SomeObject'
    ).equalTo('user', request.user);
```

The cloud function then returns this as error:

```json
{"code":107,"error":"You cannot use [object Object] as a query parameter."}
```

Stacktrace:
```
{"code":107,"stack":"Error: You cannot use [object Object] as a query parameter.
at transformQueryKeyValue (/app-path/node_modules/parse-server/lib/Adapters/Storage/Mongo/MongoTransform.js:463:11)
at transformWhere (/app-path/node_modules/parse-server/lib/Adapters/Storage/Mongo/MongoTransform.js:474:17)
at MongoStorageAdapter.find (/app-path/node_modules/parse-server/lib/Adapters/Storage/Mongo/MongoStorageAdapter.js:543:59)
at /app-path/node_modules/parse-server/lib/Controllers/DatabaseController.js:1155:33
at processTicksAndRejections (node:internal/process/task_queues:96:5)"}
```

**Comment by dblythy** [comment]
I have tried to replicate this issue with a failing test, however I can't get it to fail:

```js
// js:
it('can handle current user from cloud', async () => {
    const user = await Parse.User.signUp('bob', 'password');
    const obj = new Parse.Object('SomeObject');
    const relation = obj.relation('users');
    relation.add(user);
    await obj.save(null, {useMasterKey: true});
    const response = await Parse.Cloud.run('getObjects', null, {sessionToken: user.getSessionToken()});
    expect(response.length).toBe(1)
})
// cloud function:
Parse.Cloud.define('getObjects', (request) => {
  return new Parse.Query('SomeObject').equalTo('users', request.user).find({useMasterKey: true});
});
```

**Comment by stephannielsen** [comment]
Thank you @dblythy for looking into it. I also investigated a bit last night and it turns out to be somehow related to our parse-server package. I started with a fresh parse-server app to replicate the issue but it works fine. I am now slowly adding our other dependencies and rebuild our structure to get a minimal repo replicating the issue. The `equalTo` lookup also works in Dashboard JS console.

As the change from 3.4.1 to 3.4.2 is just tests (should have no affect) and some babel dependency updates, I am currently suspecting a dependency clash between some dependencies in our project.

The other main difference is that we do not use Parse Sessions, so `request.user` has no active session behind it...but as it works <= 3.4.1, I don't see why that suddenly would be an issue.

I notice now that I forgot to mention that we use TypeScript in our project. Added it to the description.

**Comment by dblythy** [comment]
> The other main difference is that we do not use Parse Sessions, so request.user has no active session behind it.

This could be a contributing factor. How does the cloud function infer `request.user`? Normally it does it via the header of `X-Parse-Session-Token`, which looks up the session.

Is `request.user` an instance of `Parse.Object`? It seems to me that the `encode` function isn't properly converting `request.user` to a pointer, which happens here:

https://github.com/parse-community/Parse-SDK-JS/blob/alpha/src/encode.js#L31-L43

Another thing you could try is using `equalTo('user', request.user.toPointer())`

**Comment by dblythy** [comment]
I have opened [this](https://github.com/parse-community/parse-server/pull/8164) PR that should help you debug why the error is being thrown.

**Comment by stephannielsen** [comment]
We use JWT tokens for authentication and the user is set before as `req['userFromJWT']` in our express middleware. I do see the user details in the request, so that seems to work. It is also inferred correctly as `ParseUser` / `Parse.Object` as far as I can see but I will check that again.

I will also try with `toPointer()`. That would be better then creating a new Parse.User object and assigning the id manually as a workaround.

**Comment by mtrezza** [comment]
@stephannielsen You are mentioning a Parse Server version for which we do not provide official support anymore (`3.x`). I would just like to clarify that you are experiencing the issue with the latest alpha version of Parse Server (`5.3.0-alpha.21`), is that correct?

**Comment by dblythy** [comment]
Very nice! Hopefully JWT is supported OOTB soon. 

If you are still stuck, see if you can stringify `[object Object]` and we can work out why the JSON isn't properly converted.

**Comment by dblythy** [comment]
Very nice, hopefully we get JWT support soon. 

If you are still stuck, see if you can stringify the `[object Object]` so we can diagnose why the JSON isn't being correctly encoded as a pointer.

**Comment by stephannielsen** [comment]
@mtrezza I think you misread that. The problem appeared with the JS SDK v3.4.2. Older versions of the SDK work fine. I am using parse-server 5.2.5, so the latest stable release. I will check later if the issue persists with the alpha release for parse-server, but so far it seems to be a JS SDK issue.

@dblythy Thanks. I will investigate some more. Besides this, JWT token support works fine. With #1534 I plan to add some functionality to ease the token usage in Cloud code a bit, but currently I am blocked by this problem unfortunately.

**Comment by mtrezza** [comment]
Got it, thanks for clarifying.

**Comment by stephannielsen** [comment]
> Another thing you could try is using equalTo('user', request.user.toPointer())

This works. Added it to the ticket as another workaround.

Stringifying `request.user` gives me:

```json
{"username":"385953c4-6c04-40cc-8d9d-f737028916d1","emailVerified":true,"createdAt":"2022-06-08T18:40:33.462Z","updatedAt":"2022-09-13T15:30:31.717Z","email":"","ACL":{"lIvfN5EABH":{"read":true,"write":true}},"password":"********","objectId":"lIvfN5EABH"}
```

With 3.4.1, this the stringify output:
```json
{"username":"385953c4-6c04-40cc-8d9d-f737028916d1","emailVerified":true,"createdAt":"2022-06-08T18:40:33.462Z","updatedAt":"2022-09-13T15:34:01.844Z","email":"","ACL":{"lIvfN5EABH":{"read":true,"write":true}},"password":"********","objectId":"lIvfN5EABH"}
```

Identical, except for `updatedAt` value.

**Comment by dblythy** [comment]
I can't even get the tests to fail using:

```js
const objs = await new Parse.Query('SomeObject').equalTo('users', JSON.parse(`{"username":"385953c4-6c04-40cc-8d9d-f737028916d1","emailVerified":true,"createdAt":"2022-06-08T18:40:33.462Z","updatedAt":"2022-09-13T15:30:31.717Z","email":"","ACL":{"lIvfN5EABH":{"read":true,"write":true}},"password":"********","objectId":"lIvfN5EABH"}`)).find({useMasterKey: true});
```

Is your `Parse.User` subclassed at all?

**Comment by stephannielsen** [comment]
So I got around to create a simple Parse Server app with a cloud function to reproduce this issue: https://github.com/stephannielsen/parse-js-bug-1534

Simply switch to an older version of the SDK to get it working.

When using `req['userFromJwt']`, the cloud function fails. If you use a session token via header, the cloud function succeeds. Something seems to get lost when using `userFromJwt`.

@dblythy Any ideas?

> Is your Parse.User subclassed at all?

How do you mean that? In our original repo we specify a custom User schema. In my linked test I didn't, so this does not seem to be an issue here.

**Comment by stephannielsen** [comment]
> I would just like to clarify that you are experiencing the issue with the latest alpha version of Parse Server (5.3.0-alpha.21), is that correct?

I also tested now with latest alpha versions, both SDK and Server. Issue remains.

```
    "parse": "3.5.0-alpha.3",
    "parse-server": "5.3.0-alpha.26"
```

**Comment by dblythy** [comment]
I haven't been able to replicate - see https://github.com/stephannielsen/parse-js-bug-1534/pull/1

**Comment by mtrezza** [comment]
@stephannielsen can you share a full repo where we can replicate the issue?

**Comment by mtrezza** [comment]
@stephannielsen Could you please take a look at https://github.com/stephannielsen/parse-js-bug-1534/pull/1#issuecomment-1250187869 and confirm whether the issue is resolved on your side? If there is no further response, we'll close this a non-reproducible.

**Comment by stephannielsen** [comment]
I haven't had the time to dig deeper into it. The issue is however not resolved and still persists. I have to evaluate the changes by @dblythy to my repo and understand the differences first to say more.

**Comment by dblythy** [comment]
> My changes here are:
>
> npm start automatically creates a real user with id, a relation with that user, and then runs the script script.js, which calls the cloud function with headers.

these changes were made to attempt to replicate the bug by automatically creating all the associated objects, so that all that is required to see the broken functionality, you just need to type `npm start`. The changes automatically build the situation that you’ve described fails for you, but, I can’t get it to fail.

**Comment by stephannielsen** [comment]
No, actually, with the changes to automate it, it works for me too, but I haven't had the time to dig into it why and if I can break it again. There are a few other differences: npm vs yarn or using Parse Server config from the npm package directly instead of manual configuration for example.

**Comment by mtrezza** [comment]
Closing this as unable to reproduce. Once you have any new insight please let us know and we may reopen the issue.

**Comment by stephannielsen** [comment]
Worked on this again and my repo still shows the bug in a reproducible (and semi-automated) way now. However, I found a workaround that's easy to implement.

When using `userFromJWT` on `req`, the bug is fixed by using `.toPointer()` when assigning it.

```js
  req['userFromJWT'] = parseUser?.toPointer();
```

**Comment by mtrezza** [comment]
Is this a bug? If so, we should fix it.

**Comment by stephannielsen** [comment]
As the whole feature is not really documented, you can probably can declare it as you like :) It worked before without it and now it is necessary. But maybe the bug was that it worked before.

The reason is still unclear to me and I lack knowledge about the internals of the SDK to judge on this.

As this one line fixes it for us in a single place, I don't mind so much. As the docs PR is still open, I could add a comment to include `.toPointer()` here: https://github.com/parse-community/docs/pull/819/files

**Comment by stephannielsen** [comment]
Unfortunately, we discovered now other places where we have trouble with pointers, completely unrelated to the JWT authentication mechanism.

We have a helper function like this to create pointers:
```ts
export function asPointer<T extends Parse.Attributes>(table: string, objectId: string) {
  const pointer = new Parse.Object(table) as Parse.Object<T>;
  pointer.id = objectId;
  return pointer;
}

// and use like this
const objectA = asPointer<TypeAttributes>("TypeA", "someId");
someObjectB.set("pointerProp", objectA);
```

So we don't have to write

```ts
const objectA = new Parse.Object("TypeA") as Parse.Object<TypeAttributes>;
objectA.id = "someId";

someObjectB.set("pointerProp", objectA);
```

It's not much but it worked fine like this up to JS SDK 3.4.1. Now, the code using the helper function does not work anymore. The second way works fine. Stringifying the object shows exactly the same (only `_objCount` differs). Using `objectA.toPointer()` does not help this time.

I am trying to add a scenario to my test repo but so far failed with that...

Some questions:

* Is there a difference between `new Parse.Object("Something")` and `Parse.Object.extend("Something")`?
* Looking at the source code, `toPointer()` sets `__type` to `'Pointer'` while `new Parse.Object` sets `__type` to `'Object'` - what's the difference here? A lot of the documentation uses the approach above to create pointers before saving a pointer relation...

Anyhelp appreciated. This still feels like a bug / regression in the JS SDK to me to be honest. Not sure why we are hitting these bumps now more or less alone 😦

**Comment by stephannielsen** [comment]
As there was not a big change between 3.4.1 and 3.4.2 (https://github.com/parse-community/Parse-SDK-JS/compare/3.4.1...3.4.2), I checked the resolved dependencies in more depth now.

Turns out that we had 3 Parse packages installed:
 * parse@3.4.4 (direct dependency)
 * parse@3.4.1 (from parse-server@3.5.7)
 * parse@3.3.0 (from @parse/push-adapter, from parse-server)

As a test, I enforced a resolution in `package.json` to version 3.4.4:

```
"resolutions": {
    "parse": "3.4.4"
  }
```

Now, our old code works without any change.  Overall I guess this is more related to a change in babel versions though. But the way the dependencies are defined in parse and parse-server cause our problem. **Is there a reason why parse specified explicit versions and not ranges?** (See here for reasons against pinning for libraries: https://the-guild.dev/blog/how-should-you-pin-dependencies-and-why#libraries))

yarn.lock:
```
parse@3.3.0:
  version "3.3.0"
  resolved "https://registry.yarnpkg.com/parse/-/parse-3.3.0.tgz#96b9ae164a42612c0db9a5be601a772a4b4d0df0"
  integrity sha512-SQkTDupU7JQBJpYFIpO8TlQjUtjboUdkXaak57pjoC1ZVbhaiNyLsdYbrlM0B+sNYhlvcMh7zwZW48u10+zm0A==
  dependencies:
    "@babel/runtime" "7.14.6"
    "@babel/runtime-corejs3" "7.14.6"
    idb-keyval "5.0.6"
    react-native-crypto-js "1.0.0"
    uuid "3.4.0"
    ws "7.5.0"
    xmlhttprequest "1.8.0"
  optionalDependencies:
    crypto-js "4.0.0"

parse@3.4.1:
  version "3.4.1"
  resolved "https://registry.yarnpkg.com/parse/-/parse-3.4.1.tgz#b10a622ac64dc406cd97dd47756e65b7a7291643"
  integrity sha512-XTMaHfcOwAOiWLraNtPbCzR8o94ZWjOfaZDMM2jZ1ZDB5twtK1B82lPa+N197efZhcQ+QFbW/eDJsBybD48aSQ==
  dependencies:
    "@babel/runtime" "7.15.4"
    "@babel/runtime-corejs3" "7.14.7"
    idb-keyval "6.0.3"
    react-native-crypto-js "1.0.0"
    uuid "3.4.0"
    ws "7.5.1"
    xmlhttprequest "1.8.0"
  optionalDependencies:
    crypto-js "4.1.1"

parse@3.4.4:
  version "3.4.4"
  resolved "https://registry.yarnpkg.com/parse/-/parse-3.4.4.tgz#6358f2a59df4dfac6f6f76f1589d92942249fd43"
  integrity sha512-6mnFGD3FroQtCYImRXOFF9+BohGWGwaOIo7htD0HxGVooMR1Iyb69eYj+AoLr+o3sBBGCCGmNmjNjmvUIOnSaw==
  dependencies:
    "@babel/runtime" "7.18.0"
    "@babel/runtime-corejs3" "7.17.8"
    idb-keyval "6.0.3"
    react-native-crypto-js "1.0.0"
    uuid "3.4.0"
    ws "8.6.0"
    xmlhttprequest "1.8.0"
  optionalDependencies:
    crypto-js "4.1.1"
```

`yarn why parse` output:
```
❯ yarn why parse
yarn why v1.22.19
[1/4] 🤔  Why do we have the module "parse"...?
[2/4] 🚚  Initialising dependency graph...
[3/4] 🔍  Finding dependency...
[4/4] 🚡  Calculating file sizes...
=> Found "parse@3.4.4"
info Has been hoisted to "parse"
info This module exists because it's specified in "dependencies".
info Disk size without dependencies: "6.16MB"
info Disk size with unique dependencies: "9.55MB"
info Disk size with transitive dependencies: "22.55MB"
info Number of shared dependencies: 9
=> Found "parse-server#parse@3.4.1"
info This module exists because "parse-server" depends on it.
info Disk size without dependencies: "6.2MB"
info Disk size with unique dependencies: "9.59MB"
info Disk size with transitive dependencies: "22.6MB"
info Number of shared dependencies: 9
=> Found "@parse/push-adapter#parse@3.3.0"
info This module exists because "parse-server#@parse#push-adapter" depends on it.
info Disk size without dependencies: "5.85MB"
info Disk size with unique dependencies: "9.24MB"
info Disk size with transitive dependencies: "22.25MB"
info Number of shared dependencies: 9
✨  Done in 0.51s.
```

**Comment by mtrezza** [comment]
> Is there a reason why parse specified explicit versions and not ranges?

We work with discrete dependency versions, that's why.

**Comment by stephannielsen** [comment]
Okay.

I guess it is then just recommended to use the same SDK version as parse-server defines, so for the current stable release of parse-server it is SDK 3.4.1 and when server 5.3.0 hits, we can upgrade to SDK 3.4.2. Good to know.

**Comment by mtrezza** [comment]
Yes, because we are testing Parse Server with a specific SDK version. Not every Parse JS SDK version is compatible with every version of Parse Server. Historically, we don't track that compatibility because of tests. Once we move to a monorepo with Parse Server and Parse JS SDK combined, things will be come more transparent.
